src/linker/utilities/spark.smk

```diff
@@ -0,0 +1,199 @@
+rule terminate_spark:
+    input:
+        config["results_dir"] + "/result.parquet",
+    output:
+        temp(config["results_dir"] + "/spark_logs/spark_master_terminated.txt"),
+    localrule: True
+    shell:
+        """
+        touch {output}
+        """
+
+
+rule start_spark_master:
+    output:
+        config["results_dir"] + "/spark_logs/spark_master_log.txt",
+    container:
+        "/mnt/team/simulation_science/priv/engineering/er_ecosystem/images/spark_cluster.sif"
+    params:
+        terminate_file_name=rules.terminate_spark.output,
+    shell:
+        """
+    echo "Starting spark master - logging to {output}"
+    SPARK_ROOT=/opt/spark
+    SPARK_MASTER_PORT=28508
+    SPARK_MASTER_WEBUI_PORT=28509
+    SPARK_MASTER_HOST=$(hostname -f)
+
+
+    rm -f {output} || true
+
+    $SPARK_ROOT/bin/spark-class org.apache.spark.deploy.master.Master \
+    --host $SPARK_MASTER_HOST \
+    --port $SPARK_MASTER_PORT \
+    --webui-port $SPARK_MASTER_WEBUI_PORT \
+    &> {output} &
+    spid=$!
+
+    trap "kill -9 $spid &>/dev/null" EXIT
+
+    while true; do
+
+        if ! kill -0 $spid >/dev/null 2>&1; then
```

Comment:
Can you explain this bit?

Reply:
This bit just checks if the process for the spark master is still running, and if it is not (that is, if something other than our termination command kills it), it logs to stdout that the master died under mysterious circumstances